### PR TITLE
Added a regex based db and table name storage selector

### PR DESCRIPTION
**Pull request by vikrambohra**
## Summary

Introduced an implementation of Storage Selector that accepts a regex and storage-type and selects the provided storage type if db and table match the regex pattern. if pattern does not match default storage is used. 

## Changes

- [ ] Client-facing API Changes
- [ ] Internal API Changes
- [ ] Bug Fixes
- [x] New Features
- [ ] Performance Improvements
- [ ] Code Style
- [ ] Refactoring
- [ ] Documentation
- [x] Tests

For all the boxes checked, please include additional details of the changes made in this pull request.  

## Testing Done

- [x] Manually Tested on local docker setup. Please include commands ran, and their output.
- [x] Added new tests for the changes made.
- [x] Updated existing tests to reflect the changes made.
- [ ] No tests added or updated. Please explain why. If unsure, please feel free to ask for help.
- [ ] Some other form of testing like staging or soak time in production. Please explain.

For all the boxes checked, include a detailed description of the testing done for the changes made in this pull request.
* Ran oh-s3-spark docker-compose with 2 configured storages (s3 & hdfs) with default as s3 and hdfs for tables matching the regex ".*prod.*\.table.*"

Screenshot 2024-09-14 at 9 41 02 AM

* Created 2 tables 
    * "database.table1" that doesnt match regex and defaults to s3 storage
    * "prod_database.table1" that matches regex and chooses hdfs as storage
Screenshot 2024-09-14 at 9 42 01 AM

* Verified selected storages in hts using mysql
Screenshot 2024-09-14 at 9 42 30 AM

# Additional Information

- [ ] Breaking Changes
- [ ] Deprecations
- [ ] Large PR broken into smaller PRs, and PR plan linked in the description.

For all the boxes checked, include additional details of the changes made in this pull request.

### CODE REVIEW

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/StorageSelectorConfig.java [hunk 1]
```diff
@@ -21,7 +22,7 @@ public class StorageSelectorConfig {
 
   @Autowired StorageProperties storageProperties;
 
-  @Autowired List<StorageSelector> storageSelectors;
+  @Lazy @Autowired List<StorageSelector> storageSelectors;
```
Comment (HotSushi) [inline cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/StorageSelectorConfig.java:25]: why was this required to be lazy? `provideStorageSelector` will get called during boottime, its unlikely @lazy would help much.

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelector.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import javax.annotation.PostConstruct;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+/**
+ * An implementation of {@link StorageSelector} that takes a regex and storage-type provided in the
+ * storage selector params in the yaml configuration and returns storage that's provided if the db
+ * and table name matches the regex, else returns the cluster storage default
+ */
```
Comment (HotSushi) [inline cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelector.java:21]: can you add an example of configuration file looks like for setting this up.. 
ex:
```
storages:
   storage-selector:
          a:
          b:
          c:
```

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelector.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+@Component
+@Slf4j
+public class StorageInNameRegexSelector extends BaseStorageSelector {
```
Comment (HotSushi) [inline cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelector.java:24]: `RegexStorageSelector`?

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelector.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+/**
+ * An implementation of {@link StorageSelector} that takes a regex and storage-type provided in the
+ * storage selector params in the yaml configuration and returns storage that's provided if the db
+ * and table name matches the regex, else returns the cluster storage default
+ */
+@Component
+@Slf4j
```
Comment (HotSushi) [inline cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelector.java:23]: Is there a way to mark this selector @Lazy? IF the cluster.yaml is not configured to use this selector, and if this gets initialized, `init()` will throw exception.

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/StorageSelectorConfig.java (hunk 1 above)
Comment (HotSushi) [reply cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/StorageSelectorConfig.java:25]: I see why this @lazy might've needed, but I've feeling this will error out. Can we fix the lazy loading on the selector side?

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelector.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.regex.Pattern;
+import javax.annotation.PostConstruct;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+/**
+ * An implementation of {@link StorageSelector} that takes a regex and storage-type provided in the
+ * storage selector params in the yaml configuration and returns storage that's provided if the db
+ * and table name matches the regex, else returns the cluster storage default
```
Comment (HotSushi) [inline cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelector.java:20]: Also add documentation on assumption of merging `db`+'.' +`table` and uses that for matching. Code readers will use this documentation to get the whole idea about the selector, we should include examples/assumptions/anything else thats needed etc.

File: cluster/storage/src/test/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelectorTest.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
+package com.linkedin.openhouse.cluster.storage.selector.impl;
+
```
Comment (HotSushi) [inline cluster/storage/src/test/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelectorTest.java:2]: can we add a e2e test which picks up certain cluster.yaml and we create 2 tables in there with different storages (file: vs hdfs)? and we can validate based on the get response?

**Review by HotSushi — CHANGES_REQUESTED** [review]
thanks for the unit tests, was the docker test also run with custom cluster.yaml?

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelector.java (hunk 5 above)
Comment (vikrambohra) [reply cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelector.java]: done

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelector.java (hunk 2 above)
Comment (vikrambohra) [reply cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelector.java:33]: done

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelector.java (hunk 3 above)
Comment (vikrambohra) [reply cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelector.java:37]: I think this naming is more explicit.

File: cluster/storage/src/test/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelectorTest.java (hunk 6 above)
Comment (vikrambohra) [reply cluster/storage/src/test/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelectorTest.java:2]: StorageSelectorConfigTest.testProvideStorageInNameRegexSelector() already tests the different storage scenario. I will also perform the different storage scenario in docker and update the testing done section.

Comment (HotSushi) [reply cluster/storage/src/test/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelectorTest.java:2]: TY!  Unit tests for the added modules are good! thanks for that. I meant an e2e unit test with inmemory H2 backend for example:  com.linkedin.openhouse.tables.e2e.h2.TablesControllerTest. 
This will make reproducing errors/bugs easier and increase stability of the selector code.

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelector.java (hunk 3 above)
Comment (HotSushi) [reply cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelector.java:39]: `StorageInName` might give an incorrect idea of the implementation. I interpret it as `storage whose name is specified in the tablename gets selected`: for example,
dbcaspian.table -> caspian
db.tablehdfs -> hdfs
^ this was the earlier plan, but regex is better

Comment (vikrambohra) [reply cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelector.java]: fair. changed it.

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelector.java (hunk 4 above)
Comment (vikrambohra) [reply cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelector.java]: Changed to use name based validation not requiring lazy.

File: cluster/storage/src/test/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelectorTest.java (hunk 6 above)
Comment (vikrambohra) [reply cluster/storage/src/test/java/com/linkedin/openhouse/cluster/storage/selector/impl/StorageInNameRegexSelectorTest.java]: Can we add the e2e test as part of implementing the FeatureToggleStorageSelector which we intend to use in production for migrations? 

I've replicated the e2e test with multiple storages in docker-compose and added to the testing section.

**Review by HotSushi — APPROVED** [review]

**Review by HotSushi — APPROVED** [review]
LGTM! ty for the nice unit tests
